### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by AlessioVerardo**
Hello, 
Here is the great refactoring according to the code review made by the TAs. 
The main  new features are : 
- A new Database hierarchy that works and is tested
- A new abstraction for the Login instead of having the firebase code directly in the code
- Several small changes 

Feel free to leave a comments, 
Cheers

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -51,8 +51,11 @@ dependencies {
     implementation 'androidx.navigation:navigation-ui-ktx:2.3.3'
     implementation 'androidx.gridlayout:gridlayout:1.0.0'
 
+
     testImplementation 'junit:junit:4.13.2'
     testImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+    testImplementation "org.mockito.kotlin:mockito-kotlin:3.1.0"
+    androidTestImplementation "org.mockito.kotlin:mockito-kotlin:3.1.0"
```
Comment (Rygone) [inline app/build.gradle:58]: Perfect

File: app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseConstant.kt [hunk 2]
```diff
@@ -4,71 +4,68 @@ package com.github.sdpteam15.polyevents.database
  * Constants of collection names and attribute names.
  */
 object DatabaseConstant {
```
Comment (Rygone) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseConstant.kt:6]: Good idea to use `enum class`

File: app/src/main/java/com/github/sdpteam15/polyevents/database/objects/MaterialRequestDatabaseInterface.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * @param request the request we want to add in the database
+     * @param userAccess the user profile to use its permission
+     * @return An observer that will be set to true if the communication with the DB is over and no error
+     */
+    fun createMaterialRequest(
+        request: MaterialRequest,
+        userAccess: UserProfile? = currentProfile
+    )*/
```
Comment (Rygone) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/objects/MaterialRequestDatabaseInterface.kt:49]: It maybe not a good idea to add method in comment

File: app/src/main/java/com/github/sdpteam15/polyevents/login/UserLogin.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.github.sdpteam15.polyevents.login
+
+import com.google.firebase.auth.AuthResult
+
+
+object UserLogin {
```
Comment (Rygone) [inline app/src/main/java/com/github/sdpteam15/polyevents/login/UserLogin.kt:6]: Good abstraction

File: app/src/main/java/com/github/sdpteam15/polyevents/login/UserLoginInterface.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package com.github.sdpteam15.polyevents.login
+
+import android.app.Activity
+import android.content.Intent
+import com.github.sdpteam15.polyevents.fragments.LoginFragment
+import com.github.sdpteam15.polyevents.model.UserEntity
+import com.google.android.gms.tasks.Task
+
+interface UserLoginInterface<T> {
```
Comment (Rygone) [inline app/src/main/java/com/github/sdpteam15/polyevents/login/UserLoginInterface.kt:9]: Good abstraction.

**Review by Rygone — APPROVED** [review]

File: app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt [hunk 6]
```diff
@@ -1,192 +1,81 @@
...
+        get() = field ?: MaterialRequestDatabaseFirestore
+
+    override val currentUserObservable = Observable<UserEntity>()
+    private var loadSuccess = false
+    override var currentUser: UserEntity?
+        get() {
+            if(UserLogin.currentUserLogin.isConnected()){
+                if(!loadSuccess) {
+                    currentUserObservable.postValue(UserLogin.currentUserLogin.getCurrentUser()!!,this)
+                    firestore!!.collection(USER_COLLECTION.toString())
```
Comment (Rygone) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt:52]: `USER_COLLECTION.value` not `USER_COLLECTION.toString()`

File: app/src/main/java/com/github/sdpteam15/polyevents/util/ProfileAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        PROFILE_USERS.value to element.users
+    )
+
+    override fun fromDocument(document: MutableMap<String, Any?>, id: String) = UserProfile(
+        pid = id as String?,
+        profileName = document[PROFILE_NAME] as String,
+        userRole = if ((document[PROFILE_RANK] as String) != "")
+            UserRole.valueOf(document[PROFILE_RANK] as String)
+        else UserRole.PARTICIPANT,
+        users = (document[PROFILE_USERS.value] as List<String>).toMutableList()
```
Comment (Rygone) [inline app/src/main/java/com/github/sdpteam15/polyevents/util/ProfileAdapter.kt:27]: Don't forget the ` .value`

Comment (AlessioVerardo) [reply app/src/main/java/com/github/sdpteam15/polyevents/util/ProfileAdapter.kt:27]: You're correct, I will change this :)

File: app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt (hunk 6 above)
Comment (AlessioVerardo) [reply app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt:52]: Yes, at first, I was using .toString() then I thought it was better to use .value. Thanks for poiting this out, I'll change it

File: app/src/test/java/com/github/sdpteam15/polyevents/util/ItemEntityAdapterTest.kt [hunk 8]
```diff
@@ -45,11 +45,30 @@ class ItemEntityAdapterTest {
...
+        document = ItemEntityAdapter.toItemDocument(itemEntity2, itemQuantity)
+
+        assertEquals(
+            ItemType.valueOf(
+                document[ITEM_TYPE.value] as String
+            ), itemEntity.itemType
+        )
+        assertEquals(document[ITEM_COUNT.value], itemQuantity)
+        assertEquals(document[ITEM_NAME.value], itemName)
+        assert(!document.containsKey(ITEM_DOCUMENT_ID.value))
```
Comment (Cell00phane) [inline app/src/test/java/com/github/sdpteam15/polyevents/util/ItemEntityAdapterTest.kt:72]: What's this extra test for?

File: app/src/test/java/com/github/sdpteam15/polyevents/util/UserAdapterTest.kt [hunk 9]
```diff
@@ -28,58 +29,61 @@ class UserAdapterTest {
...
-        val document = UserAdapter.toUserDocument(user)
-
-        assertEquals(document[DatabaseConstant.USER_UID], googleId)
-        assertEquals(document[DatabaseConstant.USER_USERNAME], username)
-        assertEquals(document[DatabaseConstant.USER_BIRTH_DATE], birthDate.atStartOfDay())
-        assertEquals(document[DatabaseConstant.USER_NAME], name)
-        assertEquals(document[DatabaseConstant.USER_EMAIL], email)
-        assertEquals(document[DatabaseConstant.USER_PHONE], telephone)
+        val document = UserAdapter.toDocument(user)
+        println(document)
```
Comment (Cell00phane) [inline app/src/test/java/com/github/sdpteam15/polyevents/util/UserAdapterTest.kt:40]: Maybe remove this?

**Review by Cell00phane — APPROVED** [review]
Nothing to say. This was a huge PR, but it was great work refactoring. Approved 👍 
Just not sure whether Robin's PR will build upon this one, or not because some of the changes I've already reviewed on Robin's branch.

**Pull request merged**